core/src/test/scala/weaponregex/parser/ParserTest.scala

```diff
@@ -17,4 +17,691 @@ abstract class ParserTest extends munit.FunSuite {
       case _                                    => false
     })
   }
+
+  test("Parse concat of characters") {
+    val pattern = "hello"
+    val parsedTree = Parser(pattern, parserFlavor).get
+    assert(clue(parsedTree).isInstanceOf[Concat])
+
+    (pattern zip parsedTree.children) foreach { case (char, child) =>
+      assert(clue(child) match {
+        case Character(c, _) => c == char
+        case _               => false
+      })
+    }
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  test("Parse `}` character next to long quantifier") {
+    val pattern = "a{1}}"
+    val parsedTree = Parser(pattern, parserFlavor).get
+
+    assert(clue(parsedTree).isInstanceOf[Concat])
+    assert(parsedTree.children.head.isInstanceOf[Quantifier])
+    assert(parsedTree.children.last match {
+      case Character('}', _) => true
+      case _                 => false
+    })
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  test("Parse `]` character next to character class") {
+    val pattern = "[abc]]"
+    val parsedTree = Parser(pattern, parserFlavor).get
+
+    assert(clue(parsedTree).isInstanceOf[Concat])
+    assert(parsedTree.children.head.isInstanceOf[CharacterClass])
+    assert(parsedTree.children.last match {
+      case Character(']', _) => true
+      case _                 => false
+    })
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  test("Parse or of characters") {
+    val pattern = "h|e|l|l|o"
+    val parsedTree = Parser(pattern, parserFlavor).get
+    assert(clue(parsedTree).isInstanceOf[Or])
+
+    (pattern.replace("|", "") zip parsedTree.children) foreach { case (char, child) =>
+      assert(clue(child) match {
+        case Character(c, _) => c == char
+        case _               => false
+      })
+    }
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  test("Parse or of characters with null children") {
+    val pattern = "|h|e||l|||l|o|"
+    val parsedTree = Parser(pattern, parserFlavor).get
+    assert(clue(parsedTree).isInstanceOf[Or])
+
+    assert(clue(parsedTree.children) match {
+      case Seq(
+            Empty(_),
+            Character('h', _),
+            Character('e', _),
+            Empty(_),
+            Character('l', _),
+            Empty(_),
+            Empty(_),
+            Character('l', _),
+            Character('o', _),
+            Empty(_)
+          ) =>
+        true
+      case _ => false
+    })
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  test("Parse BOL and EOL") {
+    val pattern = "^hello$"
+    val parsedTree = Parser(pattern, parserFlavor).get
+    assert(clue(parsedTree).isInstanceOf[Concat])
+    assert(clue(parsedTree.children.head).isInstanceOf[BOL])
+    assert(clue(parsedTree.children.last).isInstanceOf[EOL])
+
+    treeBuildTest(parsedTree, pattern)
+  }
+
+  val boundaryMetacharacters: String
```

Comment:
Can you move all abstract `val`s to the top? It makes it a bit more organized

Reply:
My bad 😅 Should have done it in the last commits